testdata/p4_16_samples/checksum-l4-bmv2.p4

```diff
@@ -0,0 +1,468 @@
+                hdr.tcp.res,
+                hdr.tcp.ecn,
+                hdr.tcp.ctrl,
+                hdr.tcp.window,
+                hdr.tcp.urgentPtr,
+                hdr.tcp_opt_ss,
+                hdr.tcp_opt_s,
+                hdr.tcp_opt_sack,
+                hdr.tcp_opt_nop,
+                hdr.tcp_opt_end
+            },
+            hdr.tcp.checksum, HashAlgorithm.csum16);
+        verify_checksum_with_payload(hdr.udp.isValid(),
+            { hdr.ipv4.srcAddr,
+                hdr.ipv4.dstAddr,
+                8w0,
+                hdr.ipv4.protocol,
+                meta.l4Len,
+                hdr.udp.srcPort,
+                hdr.udp.dstPort,
+                hdr.udp.length_
+            },
+            hdr.udp.checksum, HashAlgorithm.csum16);
+    }
+}
+
+control uc(inout headers hdr,
+           inout metadata meta)
+{
+    apply {
+        update_checksum(true,
+            { hdr.ipv4.version,
+                hdr.ipv4.ihl,
+                hdr.ipv4.diffserv,
+                hdr.ipv4.totalLen,
+                hdr.ipv4.identification,
+                hdr.ipv4.flags,
+                hdr.ipv4.fragOffset,
+                hdr.ipv4.ttl,
+                hdr.ipv4.protocol,
+                hdr.ipv4.srcAddr,
+                hdr.ipv4.dstAddr,
+                hdr.ipv4.options
+            },
+            hdr.ipv4.hdrChecksum, HashAlgorithm.csum16);
+
+        update_checksum_with_payload(hdr.tcp.isValid(),
+            { hdr.ipv4.srcAddr,
+                hdr.ipv4.dstAddr,
+                8w0,
+                hdr.ipv4.protocol,
+                meta.l4Len,
+                hdr.tcp.srcPort,
+                hdr.tcp.dstPort,
+                hdr.tcp.seqNo,
+                hdr.tcp.ackNo,
+                hdr.tcp.dataOffset,
+                hdr.tcp.res,
+                hdr.tcp.ecn,
+                hdr.tcp.ctrl,
+                hdr.tcp.window,
+                hdr.tcp.urgentPtr,
+                hdr.tcp_opt_ss,
+                hdr.tcp_opt_s,
+                hdr.tcp_opt_sack,
+                hdr.tcp_opt_nop,
+                hdr.tcp_opt_end
+            },
+            hdr.tcp.checksum, HashAlgorithm.csum16);
+        update_checksum_with_payload(hdr.udp.isValid(),
+            { hdr.ipv4.srcAddr,
+                hdr.ipv4.dstAddr,
+                8w0,
+                hdr.ipv4.protocol,
+                meta.l4Len,
+                hdr.udp.srcPort,
+                hdr.udp.dstPort,
+                hdr.udp.length_
+            },
+            hdr.udp.checksum, HashAlgorithm.csum16);
+    }
+}
+
+control DeparserI(packet_out packet,
+                  in headers hdr)
+{
+    apply {
+        packet.emit(hdr.ethernet);
+        packet.emit(hdr.ipv4);
+        packet.emit(hdr.tcp);
+        packet.emit(hdr.tcp_opt_ss);
+        packet.emit(hdr.tcp_opt_s);
+        packet.emit(hdr.tcp_opt_sack);
+        packet.emit(hdr.tcp_opt_nop);
+        packet.emit(hdr.tcp_opt_end);
+        packet.emit(hdr.tcp_opt_padding);
```

Comment:
This code for TCP option handling is wrong in general.  The options need not appear in this order, thus if you want to emit the same packet you received, this will not do so unless the options happen to be received in this order.  Also, some options can appear more than once in the same packet, and this code does not handle that case, either.

I would recommend that rather than trying to handle the general case, which I think is maybe a bit of a stretch for P4 without full support for stacks of header_union, it might be best to simply use varbit for TCP options in this example/test program, with no parsing of which options appear within it.  Or, ignore TCP options completely.

Reply:
I wrote the TCP options code to prove verify and update checksums works for TCP options.  Parsing TCP options for all use cases and emitting options in a correct sequence was not a goal because I cannot use a header_union. 

The order is preserved and parsing same option multiple times if I setup data structs similar to parsing MPLS labels using a header stack.  However, I don't want to bother with this one either.

Let me ignore TCP options completely in this PR.